File: crates/ruff_fmt/src/format/comprehension.rs
use rome_formatter::prelude::*;
use rome_formatter::write;

use crate::context::ASTFormatContext;
use crate::cst::Comprehension;
use crate::shared_traits::AsFormat;

pub struct FormatComprehension<'a> {
    item: &'a Comprehension,
}

impl AsFormat<ASTFormatContext<'_>> for Comprehension {
    type Format<'a> = FormatComprehension<'a>;

    fn format(&self) -> Self::Format<'_> {
        FormatComprehension { item: self }
    }
}

impl Format<ASTFormatContext<'_>> for FormatComprehension<'_> {
    fn fmt(&self, f: &mut Formatter<ASTFormatContext<'_>>) -> FormatResult<()> {
        let comprehension = self.item;

        write!(f, [soft_line_break_or_space()])?;
        write!(f, [text("for ")])?;
        write!(f, [comprehension.target.format()])?;
        write!(f, [text(" in ")])?;
        write!(f, [comprehension.iter.format()])?;
        for if_clause in &comprehension.ifs {
            write!(f, [soft_line_break_or_space()])?;
            write!(f, [text("if ")])?;
            write!(f, [if_clause.format()])?;
        }

        Ok(())
    }
}

File: crates/ruff_python_parser/resources/invalid/expressions/list/comprehension.py
# Invalid target
[x for 1 in y]
[x for 'a' in y]
[x for call() in y]
[x for {a, b} in y]

# Invalid iter
[x for x in *y]
[x for x in yield y]
[x for x in yield from y]
[x for x in lambda y: y]
[**x for x in [{1: 2}]]
[*x, for x in y]
[*x, *y for x in y]

# Invalid if
[x for x in data if *y]
[x for x in data if yield y]
[x for x in data if yield from y]
[x for x in data if lambda y: y]
[*x if x else y for x in z]
[x if x else *y for x in z]
